File: max-number-of-k-sum-pairs/max-number-of-k-sum-pairs.py
class Solution(object):
    def maxOperations(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        if len(nums) == 1:
            return 0

        # create hash map of numbers provided in input array
        numMap = dict()
        for num in nums:
            if num not in numMap:
                numMap[num] = 1
            else:
                numMap[num] += 1

        numOperations = 0

        seen = set()

        for num in nums:
            otherNum = k-num
            if (num not in seen) and (otherNum in numMap):
                if num != otherNum:
                    numOperations += min(numMap[num], numMap[otherNum])
                else:
                    numOperations += numMap[num] // 2
                seen.add(num)
                seen.add(otherNum)

        return numOperations

File: max-number-of-k-sum-pairs/max-number-of-k-sum-pairs.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function(nums, k) {
    var numMap = {}

    for (const num of nums) {
        if (num in numMap) {
            numMap[num] += 1
        } else {
            numMap[num] = 1
        }
    }

    var numOperations = 0
    var seenNums = new Set()

    for (const num of nums) {
        var otherNum = k - num
        
        if (!seenNums.has(num) && numMap.hasOwnProperty(otherNum)) {
            if (otherNum === num) {
                numOperations += Math.floor(numMap[num] / 2)
            } else {
                numOperations += Math.min(numMap[num], numMap[otherNum])
            }
            seenNums.add(num)
            seenNums.add(otherNum)
        }
    }

    return numOperations
};